In [45]:
import mxnet as mx

[Mxnet 'Getting Started' guide](http://mxnet.readthedocs.io/en/latest/get_started/index.html)

### Part one

In [3]:
train = mx.io.MNISTIter(
    image      = "mnist/train-images-idx3-ubyte",
    label      = "mnist/train-labels-idx1-ubyte",
    batch_size = 128,
    data_shape = (784, ))

MXNetError: [10:43:38] src/io/local_filesys.cc:149: Check failed: allow_null  LocalFileSystem: fail to open "mnist/train-images-idx3-ubyte"

Downloaded mnist's 'train-images-idx3-ubyte' manually into mxnet/mnist/train-images-idx3-ubyte

In [4]:
train = mx.io.MNISTIter(
    image      = "mnist/train-images-idx3-ubyte",
    label      = "mnist/train-labels-idx1-ubyte",
    batch_size = 128,
    data_shape = (784, ))

MXNetError: [09:13:23] src/io/local_filesys.cc:149: Check failed: allow_null  LocalFileSystem: fail to open "mnist/train-images-idx3-ubyte"

Downloaded all three remaining files

In [5]:
train = mx.io.MNISTIter(
    image      = "mnist/train-images-idx3-ubyte",
    label      = "mnist/train-labels-idx1-ubyte",
    batch_size = 128,
    data_shape = (784, ))

MXNetError: [09:15:13] src/io/local_filesys.cc:149: Check failed: allow_null  LocalFileSystem: fail to open "mnist/train-images-idx3-ubyte"

In [3]:
train = mx.io.MNISTIter(
    image      = "/Users/diegoandai/mxnet/mnist/train-images-idx3-ubyte",
    label      = "/Users/diegoandai/mxnet/mnist/train-labels-idx1-ubyte",
    batch_size = 128,
    data_shape = (784, ))

Adding the whole path worked with the first part but this next part has now a problem

### Part two

In [7]:
val = mx.io.MNISTIter(...)

MXNetError: [09:47:44] src/io/local_filesys.cc:149: Check failed: allow_null  LocalFileSystem: fail to open "./train-images-idx3-ubyte"

In [4]:
val   = mx.io.MNISTIter(
    image      = "/Users/diegoandai/mxnet/mnist/train-images-idx3-ubyte",
    label      = "/Users/diegoandai/mxnet/mnist/train-labels-idx1-ubyte",
    batch_size = 128,
    data_shape = (784, ))

I supose I should do that, as the val is a validation set.

### Part three

In [5]:
data = mx.symbol.Variable('data')
fc1  = mx.symbol.FullyConnected(data = data, num_hidden=128)
act1 = mx.symbol.Activation(data = fc1, act_type="relu")
fc2  = mx.symbol.FullyConnected(data = act1, num_hidden = 64)
act2 = mx.symbol.Activation(data = fc2, act_type="relu")
fc3  = mx.symbol.FullyConnected(data = act2, num_hidden=10)
mlp  = mx.symbol.SoftmaxOutput(data = fc3, name = 'softmax')

### Part four

In [6]:
model = mx.model.FeedForward(
    symbol = mlp,
    num_epoch = 20,
    learning_rate = .1)
model.fit(X = train, eval_data = val)

### Part five

In [7]:
test = mx.io.MNISTIter(
    image      = "/Users/diegoandai/mxnet/mnist/train-images-idx3-ubyte",
    label      = "/Users/diegoandai/mxnet/mnist/train-labels-idx1-ubyte",
    batch_size = 128,
    data_shape = (784, ))
model.predict(X = test)

array([[  1.93979344e-09,   1.55553061e-14,   5.62773717e-10, ...,
          8.25888403e-12,   4.65762255e-08,   1.91737897e-07],
       [  1.70120351e-09,   1.55589490e-13,   4.36169706e-15, ...,
          2.89860423e-18,   4.29003831e-06,   3.22553860e-05],
       [  1.57531811e-13,   1.51601865e-12,   8.36837657e-08, ...,
          1.50828383e-15,   9.99999881e-01,   1.13517383e-10],
       ..., 
       [  4.04535143e-07,   6.00959595e-07,   5.57745761e-06, ...,
          1.16974388e-05,   4.49131512e-05,   3.58714792e-03],
       [  2.88564811e-10,   2.50645018e-13,   3.48811979e-10, ...,
          9.68607765e-07,   3.29557659e-10,   1.72167847e-06],
       [  2.30681562e-05,   5.89239248e-12,   4.56799171e-07, ...,
          9.92708147e-01,   3.00236742e-08,   7.25279888e-03]], dtype=float32)

### Trying to understand what that means

In [46]:
test = mx.io.MNISTIter(
    image      = "/Users/diegoandai/mxnet/mnist/train-images-idx3-ubyte",
    label      = "/Users/diegoandai/mxnet/mnist/train-labels-idx1-ubyte",
    batch_size = 128,
    data_shape = (784, ))
prediction = model.predict(X = test)

In [47]:
prediction.shape

(59904, 10)

So the output is 59904 results, one for each image in the test dataset. For every row (result) there are 10 probabilities, one for each digit

In [48]:
pred_last_ten = prediction [-10:]

Taking the last 10 predictions

In [49]:
for i in range(pred_last_ten.shape[0]):
    result = 'result {} was inconclusive'.format(i+1)
    for j in range(pred_last_ten.shape[1]):
        if pred_last_ten[i][j] > 0.9:
            result = 'result {} prediction is {}'.format(i+1, j)
    print(result)


result 1 prediction is 4
result 2 prediction is 3
result 3 prediction is 6
result 4 prediction is 2
result 5 prediction is 4
result 6 prediction is 4
result 7 was inconclusive
result 8 prediction is 4
result 9 prediction is 4
result 10 prediction is 7


Were there really those numbers? I took the following module from https://gist.github.com/akesling/5358964, script to show mnist images

In [72]:
import mnist_image_funct as m

In [77]:
m.read(path = '/Users/diegoandai/mxnet/mnist/')

<generator object read at 0x13c159798>

In [74]:
for i in images:
    print(i)

NameError: name 'xrange' is not defined